File: claude_api/parser.ts
/**
 * Parser for Claude API JSON stream responses with streaming support (TypeScript version)
 */

import {
  SystemInit,
  AssistantResponse,
  UserResponse,
  ResultSummary,
  ClaudeResponse,
  ClaudeResponseImpl,
  AssistantMessage,
  UserMessage,
  TextContent,
  ThinkingContent,
  ToolUseContent,
  ToolResultContent,
  Usage,
  ExtendedUsage,
  ServerToolUse,
  StreamConfig,
  DEFAULT_STREAM_CONFIG,
  MessageContent,
  UserContent
} from './models';

export class ClaudeResponseParser {
  public system_init?: SystemInit;
  public assistant_responses: AssistantResponse[] = [];
  public user_responses: UserResponse[] = [];
  public result_summary?: ResultSummary;
  public raw_lines: string[] = [];
  public events: Record<string, any>[] = [];

  constructor() {
    // Initialize parser
  }

  parse_line(line: string): SystemInit | AssistantResponse | UserResponse | ResultSummary | null {
    if (!line.trim()) {
      return null;
    }

    try {
      // Clean the line - remove any extra whitespace and non-JSON content
      const cleanLine = line.trim();
      
      // Skip lines that don't start with { (not JSON)
      if (!cleanLine.startsWith('{')) {
        return null;
      }

      // Check if line ends with } (complete JSON)
      if (!cleanLine.endsWith('}')) {
        return null;
      }

      const data = JSON.parse(cleanLine);
      this.raw_lines.push(cleanLine);
      this.events.push(data);

      if (data.type === 'system' && data.subtype === 'init') {
        this.system_init = data as SystemInit;
        return this.system_init;
      } else if (data.type === 'assistant') {
        // Parse assistant message with proper content types
        const message_data = data.message || {};
        const content_list: MessageContent[] = [];

        for (const content of message_data.content || []) {
          const content_type = content.type;

          if (content_type === 'text') {
            content_list.push(content as TextContent);
          } else if (content_type === 'thinking') {
            content_list.push(content as ThinkingContent);
          } else if (content_type === 'tool_use') {
            content_list.push(content as ToolUseContent);
          } else {
            // Skip unknown content types
            console.warn(`Unknown content type: ${content_type}`);
            continue;
          }
        }

        // Create AssistantMessage with parsed content
        const assistant_msg: AssistantMessage = {
          ...message_data,
          content: content_list
        };

        // Create AssistantResponse
        const assistant_response: AssistantResponse = {
          type: 'assistant',
          message: assistant_msg,
          parent_tool_use_id: data.parent_tool_use_id,
          session_id: data.session_id
        };

        this.assistant_responses.push(assistant_response);
        return assistant_response;
      } else if (data.type === 'user') {
        // Parse user message (tool results)
        const message_data = data.message || {};
        const content_list: UserContent[] = [];

        for (const content of message_data.content || []) {
          if (content.type === 'tool_result') {
            // Handle case where content field is a list of content objects
            const content_field = content.content || '';
            let content_text: string;

            if (Array.isArray(content_field)) {
              // Extract text from content objects
              const text_parts: string[] = [];
              for (const content_obj of content_field) {
                if (typeof content_obj === 'object' && content_obj.type === 'text') {
                  text_parts.push(content_obj.text || '');
                }
              }
              content_text = text_parts.join('');
            } else {
              content_text = String(content_field);
            }

            // Create ToolResultContent with extracted text
            const tool_result: ToolResultContent = {
              tool_use_id: content.tool_use_id || '',
              type: 'tool_result',
              content: content_text,
              is_error: content.is_error || false
            };
            content_list.push(tool_result);
          }
        }

        // Create UserMessage with parsed content
        const user_msg: UserMessage = {
          ...message_data,
          content: content_list
        };

        // Create UserResponse
        const user_response: UserResponse = {
          type: 'user',
          message: user_msg,
          parent_tool_use_id: data.parent_tool_use_id,
          session_id: data.session_id
        };

        this.user_responses.push(user_response);
        return user_response;
      } else if (data.type === 'result') {
        try {
          this.result_summary = data as ResultSummary;
          return this.result_summary;
        } catch (validation_error) {
          console.warn(`Warning: Could not parse result summary: ${validation_error}`);
          // Create a minimal result summary with available data
          this.result_summary = {
            type: 'result',
            subtype: data.subtype || 'unknown',
            is_error: data.is_error || true,
            session_id: data.session_id || '',
            duration_ms: data.duration_ms || 0,
            duration_api_ms: data.duration_api_ms || 0,
            num_turns: data.num_turns || 0,
            result: data.result || '',
            total_cost_usd: data.total_cost_usd || 0,
            usage: data.usage
          };
          return this.result_summary;
        }
      }
    } catch (error) {
      console.error(`Error parsing line: ${error}`);
      return null;
    }

    return null;
  }

  parse_stream(stream: string): ClaudeResponse {
    const lines = stream.trim().split('\n');

    for (const line of lines) {
      this.parse_line(line);
    }

    const response = new ClaudeResponseImpl();
    response.system_init = this.system_init;
    response.assistant_responses = this.assistant_responses;
    response.user_responses = this.user_responses;
    response.result_summary = this.result_summary;
    response.raw_response = stream;

    return response;
  }

  get_message_text(): string {
    const texts: string[] = [];
    for (const response of this.assistant_responses) {
      for (const content of response.message.content) {
        if (content.type === 'text') {
          texts.push((content as TextContent).text);
        }
      }
    }
    return texts.join('\n');
  }

  get_tool_uses(): ToolUseContent[] {
    const tool_uses: ToolUseContent[] = [];
    for (const response of this.assistant_responses) {
      for (const content of response.message.content) {
        if (content.type === 'tool_use') {
          tool_uses.push(content as ToolUseContent);
        }
      }
    }
    return tool_uses;
  }

  get_thinking(): string[] {
    const thoughts: string[] = [];
    for (const response of this.assistant_responses) {
      for (const content of response.message.content) {
        if (content.type === 'thinking') {
          thoughts.push((content as ThinkingContent).thinking);
        }
      }
    }
    return thoughts;
  }

  get_tool_results(): ToolResultContent[] {
    const results: ToolResultContent[] = [];
    for (const response of this.user_responses) {
      for (const content of response.message.content) {
        if (content.type === 'tool_result') {
          results.push(content as ToolResultContent);
        }
      }
    }
    return results;
  }

  get_usage_summary(): ExtendedUsage | null {
    return this.result_summary?.usage || null;
  }

  get_cost(): number | null {
    return this.result_summary?.total_cost_usd || null;
  }

  was_successful(): boolean {
    return (
      this.result_summary !== undefined &&
      !this.result_summary.is_error &&
      this.result_summary.subtype === 'success'
    );
  }

  *get_event_stream(): Generator<Record<string, any>, void, unknown> {
    for (const event of this.events) {
      yield event;
    }
  }
}

export class StreamingResponseParser {
  private config: StreamConfig;
  private buffer: string = '';
  public parser: ClaudeResponseParser;

  constructor(config?: StreamConfig) {
    this.config = config || DEFAULT_STREAM_CONFIG;
    this.parser = new ClaudeResponseParser();
  }

  *parse_chunk(chunk: string): Generator<SystemInit | AssistantResponse | UserResponse | ResultSummary | string, void, unknown> {
    this.buffer += chunk;

    while (this.buffer.includes('\n')) {
      const newlineIndex = this.buffer.indexOf('\n');
      const line = this.buffer.substring(0, newlineIndex);
      this.buffer = this.buffer.substring(newlineIndex + 1);

      const cleanLine = line.trim();
      if (cleanLine && cleanLine.startsWith('{')) {
        if (this.config.yield_raw) {
          // Yield raw JSON string
          yield cleanLine;
        } else {
          // Parse and yield structured event
          const event = this.parser.parse_line(cleanLine);
          if (event) {
            yield event;
          }
        }
      }
    }
  }

  *flush(): Generator<SystemInit | AssistantResponse | UserResponse | ResultSummary | string, void, unknown> {
    const cleanBuffer = this.buffer.trim();
    if (cleanBuffer && cleanBuffer.startsWith('{')) {
      if (this.config.yield_raw) {
        yield cleanBuffer;
      } else {
        const event = this.parser.parse_line(cleanBuffer);
        if (event) {
          yield event;
        }
      }
    }
    this.buffer = '';
  }

  get_response(): ClaudeResponse {
    const response = new ClaudeResponseImpl();
    response.system_init = this.parser.system_init;
    response.assistant_responses = this.parser.assistant_responses;
    response.user_responses = this.parser.user_responses;
    response.result_summary = this.parser.result_summary;
    response.raw_response = this.parser.raw_lines.join('\n');

    return response;
  }
}

// Convenience functions

export function parse_claude_stream(stream: string): ClaudeResponse {
  const parser = new ClaudeResponseParser();
  return parser.parse_stream(stream);
}

export function* stream_parse_generator(
  stream: Generator<string, void, unknown>,
  config?: StreamConfig
): Generator<SystemInit | AssistantResponse | UserResponse | ResultSummary | string, void, unknown> {
  const parser = new StreamingResponseParser(config);

  for (const chunk of stream) {
    yield* parser.parse_chunk(chunk);
  }

  // Flush any remaining content
  yield* parser.flush();
}

export async function* async_stream_parse_generator(
  stream: AsyncGenerator<string, void, unknown>,
  config?: StreamConfig
): AsyncGenerator<SystemInit | AssistantResponse | UserResponse | ResultSummary | string, void, unknown> {
  const parser = new StreamingResponseParser(config);

  for await (const chunk of stream) {
    for (const event of parser.parse_chunk(chunk)) {
      yield event;
    }
  }

  // Flush any remaining content
  for (const event of parser.flush()) {
    yield event;
  }
}

File: Python/claude_api/parser.py
"""Parser for Claude API JSON stream responses with streaming support."""

import json
from typing import List, Generator, Union, Optional, Dict, Any, AsyncGenerator
from .models import (
    SystemInit, AssistantResponse, UserResponse, ResultSummary, ClaudeResponse,
    AssistantMessage, UserMessage, TextContent, ThinkingContent, ToolUseContent,
    ToolResultContent, Usage, ExtendedUsage, ServerToolUse, StreamConfig
)


class ClaudeResponseParser:
    """Parses Claude API JSON stream responses."""
    
    def __init__(self):
        """Initialize the parser."""
        self.system_init: Optional[SystemInit] = None
        self.assistant_responses: List[AssistantResponse] = []
        self.user_responses: List[UserResponse] = []
        self.result_summary: Optional[ResultSummary] = None
        self.raw_lines: List[str] = []
        self.events: List[Dict[str, Any]] = []
    
    def parse_line(self, line: str) -> Optional[Union[SystemInit, AssistantResponse, UserResponse, ResultSummary]]:
        """Parse a single JSON line from the stream."""
        if not line.strip():
            return None
            
        try:
            data = json.loads(line)
            self.raw_lines.append(line)
            self.events.append(data)
            
            if data.get("type") == "system" and data.get("subtype") == "init":
                self.system_init = SystemInit(**data)
                return self.system_init
                
            elif data.get("type") == "assistant":
                # Parse assistant message with proper content types
                message_data = data.get("message", {})
                content_list = []
                
                for content in message_data.get("content", []):
                    content_type = content.get("type")
                    
                    if content_type == "text":
                        content_list.append(TextContent(**content))
                    elif content_type == "thinking":
                        content_list.append(ThinkingContent(**content))
                    elif content_type == "tool_use":
                        content_list.append(ToolUseContent(**content))
                    else:
                        # Skip unknown content types
                        print(f"Unknown content type: {content_type}")
                        continue
                
                # Create AssistantMessage with parsed content
                message_data["content"] = content_list
                assistant_msg = AssistantMessage(**message_data)
                
                # Create AssistantResponse
                response_data = {
                    "type": "assistant",
                    "message": assistant_msg,
                    "parent_tool_use_id": data.get("parent_tool_use_id"),
                    "session_id": data.get("session_id")
                }
                assistant_response = AssistantResponse(**response_data)
                self.assistant_responses.append(assistant_response)
                return assistant_response
                
            elif data.get("type") == "user":
                # Parse user message (tool results)
                message_data = data.get("message", {})
                content_list = []
                
                for content in message_data.get("content", []):
                    if content.get("type") == "tool_result":
                        # Handle case where content field is a list of content objects
                        content_field = content.get("content", "")
                        if isinstance(content_field, list):
                            # Extract text from content objects
                            text_parts = []
                            for content_obj in content_field:
                                if isinstance(content_obj, dict) and content_obj.get("type") == "text":
                                    text_parts.append(content_obj.get("text", ""))
                            content_text = "".join(text_parts)
                        else:
                            content_text = str(content_field)
                        
                        # Create ToolResultContent with extracted text
                        tool_result = ToolResultContent(
                            tool_use_id=content.get("tool_use_id", ""),
                            type="tool_result",
                            content=content_text,
                            is_error=content.get("is_error", False)
                        )
                        content_list.append(tool_result)
                
                # Create UserMessage with parsed content
                message_data["content"] = content_list
                user_msg = UserMessage(**message_data)
                
                # Create UserResponse
                response_data = {
                    "type": "user",
                    "message": user_msg,
                    "parent_tool_use_id": data.get("parent_tool_use_id"),
                    "session_id": data.get("session_id")
                }
                user_response = UserResponse(**response_data)
                self.user_responses.append(user_response)
                return user_response
                
            elif data.get("type") == "result":
                try:
                    self.result_summary = ResultSummary(**data)
                    return self.result_summary
                except Exception as validation_error:
                    print(f"Warning: Could not parse result summary: {validation_error}")
                    # Create a minimal result summary with available data
                    self.result_summary = ResultSummary(
                        type="result",
                        subtype=data.get("subtype", "unknown"),
                        is_error=data.get("is_error", True),
                        session_id=data.get("session_id", "")
                    )
                    return self.result_summary
                
        except (json.JSONDecodeError, ValueError) as e:
            print(f"Error parsing line: {e}")
            return None
    
    def parse_stream(self, stream: str) -> ClaudeResponse:
        """Parse complete stream output into ClaudeResponse."""
        lines = stream.strip().split('\n')
        
        for line in lines:
            self.parse_line(line)
        
        return ClaudeResponse(
            system_init=self.system_init,
            assistant_responses=self.assistant_responses,
            user_responses=self.user_responses,
            result_summary=self.result_summary,
            raw_response=stream
        )
    
    def get_message_text(self) -> str:
        """Extract all assistant message text."""
        texts = []
        for response in self.assistant_responses:
            for content in response.message.content:
                if isinstance(content, TextContent):
                    texts.append(content.text)
        return "\n".join(texts)
    
    def get_tool_uses(self) -> List[ToolUseContent]:
        """Extract all tool uses from assistant messages."""
        tool_uses = []
        for response in self.assistant_responses:
            for content in response.message.content:
                if isinstance(content, ToolUseContent):
                    tool_uses.append(content)
        return tool_uses
    
    def get_thinking(self) -> List[str]:
        """Extract all thinking content."""
        thoughts = []
        for response in self.assistant_responses:
            for content in response.message.content:
                if isinstance(content, ThinkingContent):
                    thoughts.append(content.thinking)
        return thoughts
    
    def get_tool_results(self) -> List[ToolResultContent]:
        """Extract all tool results from user messages."""
        results = []
        for response in self.user_responses:
            for content in response.message.content:
                if isinstance(content, ToolResultContent):
                    results.append(content)
        return results
    
    def get_usage_summary(self) -> Optional[ExtendedUsage]:
        """Get usage summary from result."""
        if self.result_summary:
            return self.result_summary.usage
        return None
    
    def get_cost(self) -> Optional[float]:
        """Get total cost in USD."""
        if self.result_summary:
            return self.result_summary.total_cost_usd
        return None
    
    def was_successful(self) -> bool:
        """Check if the request was successful."""
        return (
            self.result_summary is not None 
            and not self.result_summary.is_error
            and self.result_summary.subtype == "success"
        )
    
    def get_event_stream(self) -> Generator[Dict[str, Any], None, None]:
        """Yield parsed events as they occurred."""
        for event in self.events:
            yield event


class StreamingResponseParser:
    """Parser for real-time streaming responses."""
    
    def __init__(self, config: Optional[StreamConfig] = None):
        """Initialize streaming parser."""
        self.config = config or StreamConfig()
        self.buffer = ""
        self.parser = ClaudeResponseParser()
    
    def parse_chunk(self, chunk: str) -> Generator[Union[SystemInit, AssistantResponse, UserResponse, ResultSummary, str], None, None]:
        """Parse a chunk of streaming data and yield events."""
        self.buffer += chunk
        
        while '\n' in self.buffer:
            line, self.buffer = self.buffer.split('\n', 1)
            
            if line.strip():
                if self.config.yield_raw:
                    # Yield raw JSON string
                    yield line
                else:
                    # Parse and yield structured event
                    event = self.parser.parse_line(line)
                    if event:
                        yield event
    
    def flush(self) -> Generator[Union[SystemInit, AssistantResponse, UserResponse, ResultSummary, str], None, None]:
        """Flush any remaining buffer content."""
        if self.buffer.strip():
            if self.config.yield_raw:
                yield self.buffer
            else:
                event = self.parser.parse_line(self.buffer)
                if event:
                    yield event
        self.buffer = ""
    
    def get_response(self) -> ClaudeResponse:
        """Get the complete parsed response."""
        return ClaudeResponse(
            system_init=self.parser.system_init,
            assistant_responses=self.parser.assistant_responses,
            user_responses=self.parser.user_responses,
            result_summary=self.parser.result_summary,
            raw_response="\n".join(self.parser.raw_lines)
        )


def parse_claude_stream(stream: str) -> ClaudeResponse:
    """Convenience function to parse Claude stream output."""
    parser = ClaudeResponseParser()
    return parser.parse_stream(stream)


def stream_parse_generator(stream: Generator[str, None, None], config: Optional[StreamConfig] = None) -> Generator[Union[SystemInit, AssistantResponse, UserResponse, ResultSummary, str], None, None]:
    """Parse a stream generator and yield events in real-time."""
    parser = StreamingResponseParser(config)
    
    for chunk in stream:
        yield from parser.parse_chunk(chunk)
    
    # Flush any remaining content
    yield from parser.flush()


async def async_stream_parse_generator(
    stream: AsyncGenerator[str, None], 
    config: Optional[StreamConfig] = None
) -> AsyncGenerator[Union[SystemInit, AssistantResponse, UserResponse, ResultSummary, str], None]:
    """Parse an async stream generator and yield events in real-time."""
    parser = StreamingResponseParser(config)
    
    async for chunk in stream:
        for event in parser.parse_chunk(chunk):
            yield event
    
    # Flush any remaining content
    for event in parser.flush():
        yield event